ta/aes_crypto_ta.c

```
@@ -153,28 +156,94 @@ static TEE_Result aes_Ctr128_Encrypt(uint32_t param_types, TEE_Param params[4])
   key = params[PARAM_AES_KEY].memref.buffer;
   key_size = params[PARAM_AES_KEY].memref.size;
 
-  if (key_size == 0)
+  if (key_size == 0 || sz == 0 || outsz == 0 || sz > outsz
+      || buf == NULL || outbuf == NULL) {
+    EMSG("%s: buffer too short\n", __func__);
     return TEE_ERROR_SHORT_BUFFER;
+  }
+  /* Validate that the destination buffer is actually in secure memory.
+   * This is mandatory to protect against a malicious REE application
+   * sending a shared (non-secure) memory buffer but identifying it as
+   * a 'Secure' buffer type. */
+  res = TEE_CheckMemoryAccessRights(TEE_MEMORY_ACCESS_ANY_OWNER |
+				    TEE_MEMORY_ACCESS_WRITE | TEE_MEMORY_ACCESS_SECURE,
+				    outbuf, outsz);
+
+  if (res != TEE_SUCCESS) {
+    EMSG("%s: WARNING: output buffer is not in secure memory", __func__);
+    /* return TEE_ERROR_SECURITY;*/
```

I am not sure to understand the intention here. In the non-SDP implementation, there will be way too more errors in the decryption path.  IMO, you need to be able to enable/disable SDP at compilation time (e.g based on CFG_SECURE_DATA_PATH). 

In the SDP case, you should return an error. In the non-SDP case, remember to check that outbuf is writable like it was done originally.

---

will fix